llama_index/llm_predictor/vellum/prompt_registry.py

```diff
@@ -0,0 +1,125 @@
+from __future__ import annotations
+
+from typing import List, Any, Dict
+from uuid import uuid4
+
+from vellum.client import Vellum
+
+from llama_index import Prompt
+from llama_index.llm_predictor.vellum.types import VellumDeployment
+
+
+class VellumPromptRegistry:
+    """Registers and retrieves prompts with Vellum.
+
+    A prompt within LlamaIndex should map to a Deployment within Vellum
+    at a minimum, and may optionally map to a Sandbox as well.
+    """
+
+    def __init__(self, vellum_api_key: str) -> None:
+        self._vellum_client = Vellum(api_key=vellum_api_key)
+
+    def from_prompt(self, prompt: Prompt) -> VellumDeployment:
+        """Retrieves a prompt from Vellum or registers a new one if needed."""
+
+        deployment_id = prompt.prompt_kwargs.get("vellum_deployment_id")
+        deployment_name = prompt.prompt_kwargs.get("vellum_deployment_name")
+
+        if deployment_id or deployment_name:
+            return VellumDeployment(
+                deployment_id=deployment_id, deployment_name=deployment_name
+            )
+
+        registered_prompt = self._register_prompt(prompt)
+
+        return VellumDeployment(
+            deployment_id=registered_prompt["deployment"]["id"],
+            deployment_name=registered_prompt["deployment"]["name"],
+            sandbox_id=registered_prompt["sandbox"]["id"],
+        )
+
+    def _register_prompt(self, prompt: Prompt) -> Dict[str, Any]:
+        """Registers a prompt with Vellum.
+
+        Prompts should ideally have `vellum_label` and `vellum_name` properties
+        within prompt_kwargs set. These are used to identify the entities
+        associated with the prompt within Vellum.
+
+        By registering a prompt, Vellum will:
+        1) Create a Sandbox for the prompt.
+        2) Create a Deployment for the prompt.
+        """
+
+        label = (
+            prompt.prompt_kwargs.get("vellum_label")
+            or f"LlamaIndex Demo: {prompt.prompt_type}"
+        )
+        name = (
+            prompt.prompt_kwargs.get("vellum_name")
+            or f"llama-index-demo-{prompt.prompt_type}"
+        )
+
+        # TODO: Pass this payload to an API that creates a sandbox,
+        #  model version, and deployment.
+        payload = {  # noqa
+            "name": name,
+            "label": label,
+            "provider": "OPENAI",
+            "model": "text-davinci-003",
+            "prompt": self._contruct_prompt_data(prompt),
+        }
+
+        return {
```

Comment:
This just returns hard-coded data for now, but the idea is to hit an API with this payload.

Reply:
Is the idea to expose some of this through the VellumPredictor constructor as well? So user can at least specify LLM model?

Reply:
It's a good open Q. My understanding is that in LlamaIndex today, all prompts use `text-davinci-003` by default. This PR follows this pattern by using `text-davinci-003` by default, but then letting users swap the model through Vellum's UI (possibly as a result of using our UI to experiment with finding the best model).